src/test/py/bazel/bzlmod/bazel_module_test.py

```diff
@@ -575,15 +568,100 @@ def setUpProjectWithLocalRegistryModule(self, dep_name, dep_version,
+    # passes runfiles along
+    self.main_registry.createLocalPathModule('bare_rule', '1.0', 'bare_rule')
+    projects_dir.joinpath('bare_rule').mkdir(exist_ok=True)
+    scratchFile(projects_dir.joinpath('bare_rule', 'WORKSPACE'))
+    scratchFile(projects_dir.joinpath('bare_rule', 'BUILD'))
+    scratchFile(
+        projects_dir.joinpath('bare_rule', 'defs.bzl'), [
+            'def _bare_binary_impl(ctx):',
+            '  exe = ctx.actions.declare_file(ctx.label.name)',
+            '  ctx.actions.write(exe, "i got nothing", True)',
+            '  runfiles = ctx.runfiles(files=ctx.files.data)',
+            '  for data in ctx.attr.data:',
+            '    runfiles = runfiles.merge(data[DefaultInfo].default_runfiles)',
+            '  return DefaultInfo(files=depset(direct=[exe]), executable=exe, runfiles=runfiles)',
+            'bare_binary=rule(',
+            '  implementation=_bare_binary_impl,',
+            '  attrs={"data":attr.label_list(allow_files=True)},',
+            '  executable=True,',
+            ')',
+        ])
+
+    # Now set up a project tree shaped like a diamond
+    self.ScratchFile('MODULE.bazel', [
+        'module(name="me",version="1.0")',
+        'bazel_dep(name="foo",version="1.0")',
+        'bazel_dep(name="bar",version="2.0")',
+        'bazel_dep(name="bare_rule",version="1.0")',
+    ])
+    self.ScratchFile('WORKSPACE')
+    self.ScratchFile('WORKSPACE.bzlmod', ['workspace(name="me_ws")'])
+    self.ScratchFile('BUILD', [
+        'load("@bare_rule//:defs.bzl", "bare_binary")',
+        'bare_binary(name="me",data=["@foo"])',
+    ])
+    self.main_registry.createLocalPathModule('foo', '1.0', 'foo', {
+        'quux': '1.0',
+        'bare_rule': '1.0'
+    })
+    self.main_registry.createLocalPathModule('bar', '2.0', 'bar', {
+        'quux': '2.0',
+        'bare_rule': '1.0'
+    })
+    self.main_registry.createLocalPathModule('quux', '1.0', 'quux1',
+                                             {'bare_rule': '1.0'})
+    self.main_registry.createLocalPathModule('quux', '2.0', 'quux2',
+                                             {'bare_rule': '1.0'})
+    for dir_name, build_file in [
+        ('foo', 'bare_binary(name="foo",data=["@quux"])'),
+        ('bar', 'bare_binary(name="bar",data=["@quux"])'),
+        ('quux1', 'bare_binary(name="quux")'),
+        ('quux2', 'bare_binary(name="quux")'),
+    ]:
+      projects_dir.joinpath(dir_name).mkdir(exist_ok=True)
+      scratchFile(projects_dir.joinpath(dir_name, 'WORKSPACE'))
+      scratchFile(
+          projects_dir.joinpath(dir_name, 'BUILD'), [
+              'load("@bare_rule//:defs.bzl", "bare_binary")',
+              'package(default_visibility=["//visibility:public"])',
+              build_file,
+          ])
+
+    # Finally we get to build stuff!
+    self.RunBazel(['build', '//:me', '@bar//:bar'], allow_failure=False)
+    with open(self.Path('bazel-bin/me.repo_mapping'), 'r') as f:
```

Comment:
Could you build this check into a binary executed as part of an action? That would give use confidence that the file exists in the sandbox, not just in the output tree.

That said, we may want a test that covers remote execution.

Reply:
I made the `bare_binary` fail if it can't find the repo mapping manifest file, and changed `bazel build` to `bazel run`. The content assertion is still outside.

With the current way the repo mapping manifest file is generated, I don't think remote execution is a concern anymore -- it's just another normal artifact, and has no entanglement with the special handling around the symlink tree.

Reply:
and of course the shell script doesn't work on windows... give me a second to fix that

Reply:
Agreed on the remote execution part. By making the binary a test and running `bazel test`, you would get coverage for the sandbox case without more code.